File: android/app/src/main/java/com/phonespeaker/app/core/TcpClient.kt
package com.phonespeaker.app.core

import java.io.EOFException
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.Socket

/**
 * TcpClient.kt — 凍結共用 TCP framing 輔助（§13：wifi / u1 / u2 共用）。
 *
 * 對應 PC 端 pc/core/tcp_client.py，只負責「怎麼在一條已建立的 socket 上
 * 收送 Frame」，不知道 socket 是誰建立的、也不知道上層是哪個 transport 在用。
 *
 * 依 §10.3：這個檔案一旦 WiFi 驗收通過就視為凍結，之後 U1/U2 只能「使用」它。
 */
private const val HEADER_LEN = 4
private const val TYPE_LEN = 1
private const val MAX_FRAME_LEN = 64 * 1024 * 1024 // 64MB 上限，防止壞資料造成記憶體暴衝

class ConnectionClosedException(message: String) : IOException(message)

object TcpFraming {

    @Throws(IOException::class)
    fun sendFrame(out: OutputStream, frame: Frame) {
        val body = ByteArray(TYPE_LEN + frame.payload.size)
        body[0] = frame.type.value.toByte()
        System.arraycopy(frame.payload, 0, body, TYPE_LEN, frame.payload.size)

        val header = ByteArray(HEADER_LEN)
        writeBigEndianInt(header, 0, body.size)

        out.write(header)
        out.write(body)
        out.flush()
    }

    @Throws(IOException::class)
    fun recvFrame(input: InputStream): Frame {
        val header = readExact(input, HEADER_LEN)
        val length = readBigEndianInt(header, 0)
        if (length < TYPE_LEN || length > MAX_FRAME_LEN) {
            throw IOException("frame 長度不合法: $length")
        }
        val body = readExact(input, length)
        val typeByte = body[0].toInt() and 0xFF
        val type = MsgType.fromValue(typeByte)
        val payload = body.copyOfRange(TYPE_LEN, body.size)
        return Frame(type, payload)
    }

    @Throws(IOException::class)
    private fun readExact(input: InputStream, n: Int): ByteArray {
        val buf = ByteArray(n)
        var offset = 0
        while (offset < n) {
            val read = input.read(buf, offset, n - offset)
            if (read < 0) {
                throw ConnectionClosedException("對端關閉連線（read 回傳 -1）")
            }
            if (read == 0) {
                // 理論上 blocking socket read() 不該回傳 0；防禦性處理避免忙迴圈。
                throw EOFException("讀到 0 bytes（非預期）")
            }
            offset += read
        }
        return buf
    }

    private fun writeBigEndianInt(buf: ByteArray, offset: Int, value: Int) {
        buf[offset] = ((value ushr 24) and 0xFF).toByte()
        buf[offset + 1] = ((value ushr 16) and 0xFF).toByte()
        buf[offset + 2] = ((value ushr 8) and 0xFF).toByte()
        buf[offset + 3] = (value and 0xFF).toByte()
    }

    private fun readBigEndianInt(buf: ByteArray, offset: Int): Int {
        return ((buf[offset].toInt() and 0xFF) shl 24) or
            ((buf[offset + 1].toInt() and 0xFF) shl 16) or
            ((buf[offset + 2].toInt() and 0xFF) shl 8) or
            (buf[offset + 3].toInt() and 0xFF)
    }

    /** 低延遲導向：關掉 Nagle，避免小封包被延遲合併。 */
    @Throws(IOException::class)
    fun configureSocketForStreaming(socket: Socket) {
        socket.tcpNoDelay = true
    }
}

File: pc/core/tcp_client.py
"""tcp_client.py — 凍結共用 TCP framing 輔助（§13：wifi / u1 / u2 共用）。

只負責「怎麼在一條已建立的 socket 上收送 Frame」，完全不知道 socket 是怎麼
建立的（server accept 或 client connect），也不知道上層是誰在用它
（WiFi / U1 RNDIS / U2 adb-forward 之後都走同一條 TCP，複用同一套 framing）。

依 §10.3 檔案隔離原則：這個檔案一旦通過 WiFi 驗收就視為凍結，之後 U1/U2
只能「使用」它，不應該為了某個 transport 的特殊需求去修改它的行為。
"""

from __future__ import annotations

import logging
import socket
import struct
from typing import Optional

from .handshake import Frame, MsgType

logger = logging.getLogger(__name__)

_LENGTH_STRUCT = struct.Struct(">I")  # 4 bytes, big-endian, u32
_HEADER_LEN = _LENGTH_STRUCT.size
_TYPE_LEN = 1
_MAX_FRAME_LEN = 64 * 1024 * 1024  # 64MB 上限，防止壞資料造成記憶體暴衝


class ConnectionClosed(Exception):
    """對端關閉連線或讀到不完整資料時拋出。"""


def send_frame(sock: socket.socket, frame: Frame) -> None:
    """送出一個 frame：[4B length][1B type][payload]。"""
    body = bytes([frame.type]) + frame.payload
    header = _LENGTH_STRUCT.pack(len(body))
    sock.sendall(header + body)


def recv_exact(sock: socket.socket, n: int) -> bytes:
    """讀滿 n bytes，對端提早關閉則拋 ConnectionClosed。"""
    chunks = bytearray()
    while len(chunks) < n:
        chunk = sock.recv(n - len(chunks))
        if not chunk:
            raise ConnectionClosed("對端關閉連線（recv 回傳空資料）")
        chunks.extend(chunk)
    return bytes(chunks)


def recv_frame(sock: socket.socket) -> Frame:
    """讀一個完整 frame。長度或 type 不合法時拋 ValueError。"""
    header = recv_exact(sock, _HEADER_LEN)
    (length,) = _LENGTH_STRUCT.unpack(header)
    if length < _TYPE_LEN or length > _MAX_FRAME_LEN:
        raise ValueError(f"frame 長度不合法: {length}")
    body = recv_exact(sock, length)
    type_byte = body[0]
    try:
        msg_type = MsgType(type_byte)
    except ValueError as e:
        raise ValueError(f"未知的 frame type: {type_byte}") from e
    return Frame(msg_type, body[_TYPE_LEN:])


def configure_socket_for_streaming(sock: socket.socket) -> None:
    """低延遲導向的 socket 選項：關掉 Nagle，避免小封包被延遲合併。"""
    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)


def try_close(sock: Optional[socket.socket]) -> None:
    """盡力關閉 socket，任何錯誤都吞掉並記 log（關閉流程不該讓程式炸掉）。"""
    if sock is None:
        return
    try:
        sock.shutdown(socket.SHUT_RDWR)
    except OSError:
        pass
    try:
        sock.close()
    except OSError as e:
        logger.debug("關閉 socket 時發生非致命錯誤: %s", e)
